modules/gapi/src/compiler/passes/match_pattern_to_substitute.cpp

```diff
@@ -0,0 +1,102 @@
+// @param firstMeta - metadata of first
+// @param secondMeta - metadata of second
+bool compareDataNodes(const Metadata& firstMeta,
+                      const Metadata& secondMeta) {
+    GAPI_Assert(firstMeta.get<NodeType>().t == NodeType::DATA);
+    GAPI_Assert(firstMeta.get<NodeType>().t == secondMeta.get<NodeType>().t);
+
+    const auto& firstData = firstMeta.get<Data>();
+    const auto& secondData = secondMeta.get<Data>();
+    // compare shape
+    if (firstData.shape != secondData.shape) {
+        return false;
+    }
+    // compare storage
+    if (firstData.storage != secondData.storage) {
+        return false;
+    }
+
+    // FIXME: is it enough to only check shape && storage?
+    return true;
+}
+
+// Returns a matched pairs of {pattern node, substitute node}
+SubgraphMatch::M matchDataNodes(const Graph& pattern, const Graph& substitute,
+    const std::vector<ade::NodeHandle>& patternNodes,
+    std::vector<ade::NodeHandle> substituteNodes) {
+    SubgraphMatch::M matched;
+    for (auto it : ade::util::zip(patternNodes, substituteNodes)) {
+        const auto& pNode = std::get<0>(it);
+        const auto& sNode = std::get<1>(it);
+        if (!compareDataNodes(pattern.metadata(pNode), substitute.metadata(sNode))) {
+            return {};
+        }
+        matched.insert({ pNode, sNode });
+    }
+    return matched;
+}
+}  // anonymous namespace
+
+SubgraphMatch matchPatternToSubstitute(const cv::gimpl::GModel::Graph& pattern,
+    const cv::gimpl::GModel::Graph& substitute,
+    const cv::gimpl::Protocol& patternP,
+    const cv::gimpl::Protocol& substituteP) {
+    //---------------------------------------------------------------
+    // Match data nodes which start and end our pattern and substitute
+    const auto& patternDataInputs = patternP.in_nhs;
+    const auto& patternDataOutputs = patternP.out_nhs;
+
+    const auto& substituteDataInputs = substituteP.in_nhs;
+    const auto& substituteDataOutputs = substituteP.out_nhs;
+
+    // if number of data nodes doesn't match, abort
+    if (patternDataInputs.size() != substituteDataInputs.size()
+        || patternDataOutputs.size() != substituteDataOutputs.size()) {
+        return {};
+    }
+
+    // for each pattern input we must find a corresponding substitute input
+    auto matchedDataInputs = matchDataNodes(pattern, substitute, patternDataInputs,
+        substituteDataInputs);
+    // if nothing found, abort
+    if (matchedDataInputs.empty()) {
+        return {};
+    }
+    auto matchedDataOutputs = matchDataNodes(pattern, substitute, patternDataOutputs,
+        substituteDataOutputs);
+    // if nothing found, abort
+    if (matchedDataOutputs.empty()) {
+        return {};
+    }
+
+    //---------------------------------------------------------------
+    // Construct SubgraphMatch object
+    SubgraphMatch match;
+    match.inputDataNodes = std::move(matchedDataInputs);
+    match.outputDataNodes = std::move(matchedDataOutputs);
+    return match;
```

Comment:
New logic. This part is matching of __pattern__ to __substitute__. Indeed it is simpler than finding a pattern in a __main graph__: we just need to match inputs && outputs (the matching itself is very simple && I'm not sure how we can make it more robust). I re-used `SubgraphMatch` here (introduced before & already in master) to reduce the amount of entities "in flight"

Reply:
>  Indeed it is simpler than finding a pattern in a main graph: we just need to match inputs && outputs 

Why do you think so?

Reply:
Matching which is performed only for IN/OUT data nodes is enough because substitution logic here takes IN/OUT nodes from matched subgraph of the test graph and replace all other stuff of this matched subgraph with the substituton content.
I assume we shall not take care of substitution's content, it is on the user ownership.
If so, then we may say that such logic is kind of robust.

Reply:
it's not quite robust to me, because in this setup we rely heavily on the user and if something incorrect is passed there, we'll get the error only on `GComputation::compile()` or worse - on `.apply()`. Ideally it'd be good to detect some semantic differences early on but I'm unsure how, so decided to leave it be like this for now

Reply:
IMHO, it is good enough for now to leave the verification of the inserted substitute to the `inferMeta` and related passes (that should fail if there are any op/data inconsistencies, etc.)

Reply:
Haven't got last idea

Reply:
the idea is: if user messed up the substitute (and instead of e.g. resize we have a matmul or something -> so the complete operation changed) then later steps like `inferMeta` will hopefully fail when there's a difference: we have orig graph, then invalid substitute, then orig graph again -> the invalid substitute triggers errors in steps when data descs do not match, etc. - so in theory, if we did the substitution to an "invalid" substitute then compilation procedure should be able to see it and report an error (though, very opaque). So we have a "last level" defence when doing `compile()`

Reply:
no, this check is done separately, but I think you already saw that